dali/python/nvidia/dali/external_source.py

```diff
@@ -0,0 +1,258 @@
+            raise ValueError("`num_outputs` is not compatible with named `ExternalSource`")
+
+        self._name = name
+        self._num_outputs = num_outputs
+        self._callback = callback
+
+        self._spec.AddArg("device", device)
+        for key, value in kwargs.items():
+            self._spec.AddArg(key, value)
+
+    @property
+    def spec(self):
+        return self._spec
+
+    @property
+    def schema(self):
+        return self._schema
+
+    @property
+    def device(self):
+        return self._device
+
+    @property
+    def preserve(self):
+        return False
+
+    def __call__(self, *, source = None, cycle = None, name = None, layout = None, **kwargs):
+        """
+        `source` : callable or iterable
+        The source of the data. The source is polled for data (via a call or `next(source)` whenever
+        the pipeline needs input for the next iteration. The source can supply one or data batches,
+        depending on the value or `num_outputs`. If `num_outputs` is not set, the `source` is expected
+        to return a single batch. If it's specified, the data is expected to a tuple or list where each
+        element corresponds to respective return value of the external_source.
+
+        `cycle`: bool
+        If `True`, the source iterable will be wrapped. Otherwise, StopIteration error wil be raised
+        when end of data is reached. Setting this flag to True when `source` is not an iterable is an
+        error.
+
+        `num_outputs` : int, optional
+        If specified, denotes the number of TensorLists produced by the source function
+
+        `name` : str, optional
+        The name of the data node - used when feeding the data in `iter_setup`; can be omitted if
+        the data is provided by `source`.
+
+        `layout` : str or list/tuple of str:
+        If provided, sets the layout of the data. When `num_outputs` > 1, layout can be a list
+        containing a distinct layout for each output. If the list has fewer elements than `num_outputs`,
+        only the first outputs have the layout set, the reset have it cleared.
+        """
+        if source is None:
+            if cycle is not None:
+                raise ValueError("The argument `cycle` can only be specified if `source` is iterable")
+            callback = self._callback
+        else:
+            if self._callback is not None:
+                raise RuntimeError("`source` already specified in constructor.")
+            callback = _get_callback_from_source(source, cycle)
+
+        if layout is not None and self._layout is not None:
+            raise RuntimeError("`layout` already specified in constructor.")
+        if name is None:
+            name = self._name
+
+        if name is not None and self._num_outputs is not None:
+            raise RuntimeError("`num_outputs` is not compatible with named `ExternalSource`")
+
+        if self._num_outputs is not None:
```

Comment:
Why can't we treat None as == 1?

Reply:
When _num_outputs == 1, the source is expected to produce a one-element list (or tuple) of batches. When it's None, it's expected to return the batch directly, without extra encapsulation.